### PR TITLE
Support both Swift 3.x and Swift 4

### PR DESCRIPTION
**Pull request by ikesyo**
Supersedes #733.

### CODE REVIEW

**Comment by ikesyo** [comment]
Test codes should be updated with https://github.com/Quick/Nimble/pull/457.

**Comment by freak4pc** [comment]
@ikesyo The failing tests run just fine locally. Perhaps you should try re-trigger a Travis build?

**Comment by ikesyo** [comment]
This should be good to go now. Too many conditional `@objc` attributes are required...

File: Tests/QuickTests/QuickFocusedTests/FocusedTests.swift [hunk 1]
```diff
@@ -36,7 +36,7 @@ class _FunctionalTests_FocusedSpec_Focused: QuickSpec {
         }
 
         fitBehavesLike("two passing shared examples")
-        fitBehavesLike(FunctionalTests_FocusedSpec_Behavior.self) {_ in ()}
+        fitBehavesLike(FunctionalTests_FocusedSpec_Behavior.self) { () -> Void in }
```
Comment (ikesyo) [inline Tests/QuickTests/QuickFocusedTests/FocusedTests.swift]: Fixed the `Expression type '()' is ambiguous without more context` error.

File: Tests/QuickTests/QuickFocusedTests/FocusedTests.swift [hunk 2]
```diff
@@ -63,6 +63,6 @@ final class FocusedTests: XCTestCase, XCTestCaseProvider {
             _FunctionalTests_FocusedSpec_Focused.self,
             _FunctionalTests_FocusedSpec_Unfocused.self
         ])
-        XCTAssertEqual(result?.executionCount, 8 as UInt)
+        XCTAssertEqual(result?.executionCount, 8)
```
Comment (ikesyo) [inline Tests/QuickTests/QuickFocusedTests/FocusedTests.swift]: `executionCount` is `Int` in Swift 3.2 / Swift 4.

File: Sources/Quick/QuickTestSuite.swift [hunk 3]
```diff
@@ -36,6 +36,7 @@ public class QuickTestSuite: XCTestSuite {
      It is expected that the first call should return a valid test suite, and
      all subsequent calls should return `nil`.
      */
+    @objc
```
Comment (ikesyo) [inline Sources/Quick/QuickTestSuite.swift]: The file is entirely surrounded with `#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)`.

File: .swiftlint.yml [hunk 4]
```diff
@@ -1,9 +1,8 @@
 disabled_rules:
   - line_length
   - type_name
-  - valid_docs
   - function_body_length
-  - variable_name
+  - identifier_name
```
Comment (ikesyo) [inline .swiftlint.yml]: > Loading configuration from '.swiftlint.yml'
__'variable_name'__ rule has been renamed to __'identifier_name'__ and will be completely removed in a future release.
configuration error: 'valid_docs' is not a valid rule identifier

File: .swiftlint.yml [hunk 5]
```diff
@@ -1,9 +1,8 @@
 disabled_rules:
   - line_length
   - type_name
-  - valid_docs
```
Comment (ikesyo) [inline .swiftlint.yml]: > Loading configuration from '.swiftlint.yml'
'variable_name' rule has been renamed to 'identifier_name' and will be completely removed in a future release.
configuration error: __'valid_docs'__ is not a valid rule identifier

**Comment by ikesyo** [comment]
Please review @Quick/core 🙏

**Comment by ikesyo** [comment]
Any concerns? I'm going to merge this if there are no recations within a few days 🙏

**Comment by ikesyo** [comment]
> Did you investigate @objcMembers?

I think that we should still conditionalize that attrubute as follows which will require much more duplication:

```swift
#if ...
@objcMembers
final public class Example: NSObject {
    ...
}
#else
final public class Example: NSObject {
    ...
}
```

**Comment by ikesyo** [comment]
The simplest solution here is just setting `SWIFT_SWIFT3_OBJC_INFERENCE` to `On`.

**Comment by ikesyo** [comment]
@sharplet It appears that `@objcMembers` is not defined in Swift 3.1 so that can't be conditionalized even if we surround it with `#if swift(>=4.0)` check. Xcode 9 (Swift 4 compiler) is required to use the attribute and must drop Xcode 8.3 (Swift 3.1) support .

**Comment by sharplet** [comment]
@ikesyo It should work with `#if swift(>=3.2)`.

**Comment by ikesyo** [comment]
@sharplet I'm saying abou Xcode 8.3 (Swift 3.1) support. The attribute is not valid Swift code and does not build in Swift 3.1:

```swift
#if os(Linux)
class A {}
#else
@objcMembers
class A {}
#endif
```

http://swift.sandbox.bluemix.net/#/repl/59ade4f213bc0e791f217a9a

**Comment by ikesyo** [comment]
Somehow `#if swift(>=3.2) && (os(macOS) || os(iOS) || os(tvOS) || os(watchOS)) && !SWIFT_PACKAGE` does not work but splitting to `#if swift(>=3.2)` and `#if (os(macOS) || os(iOS) || os(tvOS) || os(watchOS)) && !SWIFT_PACKAGE` works. 🤔

File: Tests/QuickTests/QuickFocusedTests/FocusedTests.swift (hunk 2 above)
Comment (sharplet) [reply Tests/QuickTests/QuickFocusedTests/FocusedTests.swift]: Do the tests still compile in Swift 3.1? Or was this redundant?

**Comment by sharplet** [comment]
> ...and it will be really simple to remove it in the future.

Actually, I think I mean it will be _easy_ to remove. 😅

File: Tests/QuickTests/QuickFocusedTests/FocusedTests.swift (hunk 2 above)
Comment (ikesyo) [reply Tests/QuickTests/QuickFocusedTests/FocusedTests.swift]: See the build matrix, this does still compile in Swift 3.1. In that case `8` should be inferred as `UInt` I think.

**Comment by ikesyo** [comment]
Thanks for your review and the input @sharplet!

**Pull request merged**